torch/csrc/jit/script/compiler.cpp

```diff
@@ -899,12 +899,17 @@ struct to_ir {
     // Register outputs in each block
     for (const auto& x : mutated_variables) {
       auto tv = save_true->getVar(x, stmt.range());
-      true_block->registerOutput(tv);
       auto fv = save_false->getVar(x, stmt.range());
+      auto unified = unifyTypes(tv->type(), fv->type());
+      if (!unified) {
+        throw ErrorReport(stmt)
+          << "Type mismatch: " << x << " is set to type " << tv->type()->str() << " in the true branch"
```

Comment:
It should be possible for if-statments to define variables of different types in each branch as long as that variable is not already in scope:

```
if x:
  y = 4
  print(y)
else:
  y = "hi"
  print(y)
```

What happens in this case with the current code? I think it would produce a type mismatch error. Instead, it should just choose not to expose the y variable.

Reply:
Yea good point. This would throw an error. If that value is used outside of the true/false block for something other than print we would still need to throw an error right ?

Reply:
If the variable is in scope, should we be asserting that it the new type is a subtype of its previous type? That's the behavior in emitWhile, and in assignment